File: app/src/main/java/com/topjohnwu/magisk/view/MarkDownWindow.java
package com.topjohnwu.magisk.view;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import com.topjohnwu.magisk.R;
import com.topjohnwu.net.Networking;
import com.topjohnwu.net.ResponseListener;

import java.io.InputStream;
import java.util.Scanner;

import androidx.appcompat.app.AlertDialog;
import ru.noties.markwon.Markwon;
import ru.noties.markwon.html.HtmlPlugin;
import ru.noties.markwon.image.ImagesPlugin;
import ru.noties.markwon.image.svg.SvgPlugin;

public class MarkDownWindow {

    public static void show(Context activity, String title, String url) {
        Networking.get(url).getAsString(new Listener(activity, title));
    }

    public static void show(Context activity, String title, InputStream is) {
        try (Scanner s = new Scanner(is, "UTF-8")) {
            s.useDelimiter("\\A");
            new Listener(activity, title).onResponse(s.next());
        }
    }

    static class Listener implements ResponseListener<String> {

        Context activity;
        String title;

        Listener(Context a, String t) {
            activity = a;
            title = t;
        }

        @Override
        public void onResponse(String md) {
            Markwon markwon = Markwon.builder(activity)
                    .usePlugin(HtmlPlugin.create())
                    .usePlugin(ImagesPlugin.create(activity))
                    .usePlugin(SvgPlugin.create(activity.getResources()))
                    .build();
            AlertDialog.Builder alert = new AlertDialog.Builder(activity);
            alert.setTitle(title);
            View mv = LayoutInflater.from(activity).inflate(R.layout.markdown_window, null);
            TextView tv = mv.findViewById(R.id.md_txt);
            try {
                markwon.setMarkdown(tv, md);
            } catch (ExceptionInInitializerError e) {
                //Nothing we can do about this error other than show error message
                tv.setText(R.string.download_file_error);
            }
            alert.setView(mv);
            alert.setNegativeButton(R.string.close, (dialog, id) -> dialog.dismiss());
            alert.show();
        }
    }
}

File: app/src/main/java/com/topjohnwu/magisk/view/MarkDownWindow.kt
package com.topjohnwu.magisk.view

import android.content.Context
import android.view.LayoutInflater
import android.widget.TextView
import com.topjohnwu.magisk.R
import com.topjohnwu.magisk.data.repository.StringRepository
import com.topjohnwu.magisk.ktx.coroutineScope
import io.noties.markwon.Markwon
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.koin.core.KoinComponent
import org.koin.core.inject
import timber.log.Timber
import kotlin.coroutines.coroutineContext

object MarkDownWindow : KoinComponent {

    private val repo: StringRepository by inject()
    private val markwon: Markwon by inject()

    suspend fun show(activity: Context, title: String?, url: String) {
        show(activity, title) {
            repo.getString(url)
        }
    }

    suspend fun show(activity: Context, title: String?, input: suspend () -> String) {
        val view = LayoutInflater.from(activity).inflate(R.layout.markdown_window_md2, null)

        MagiskDialog(activity)
            .applyTitle(title ?: "")
            .applyView(view)
            .applyButton(MagiskDialog.ButtonType.NEGATIVE) {
                titleRes = android.R.string.cancel
            }
            .reveal()

        val tv = view.findViewById<TextView>(R.id.md_txt)
        tv.coroutineScope = CoroutineScope(coroutineContext)
        withContext(Dispatchers.IO) {
            try {
                markwon.setMarkdown(tv, input())
            } catch (e: Exception) {
                if (e is CancellationException)
                    throw e
                Timber.e(e)
                tv.post { tv.setText(R.string.download_file_error) }
            }
        }
    }
}
